src/storage/store/replication/replicator.rs

```diff
@@ -0,0 +1,306 @@
+        &self,
+        height: u64,
+        shard: u32,
+        start_fid: u64,
+        end_fid: u64,
+    ) -> Result<ReplicationTransactions, ReplicationError> {
+        let stores = match self.stores.get(height, shard) {
+            Some(stores) => stores,
+            None => {
+                return Err(ReplicationError::StoreNotFound(
+                    height,
+                    shard,
+                    "No stores found for the given height and shard".to_string(),
+                ));
+            }
+        };
+
+        // TODO: validate fid range
+
+        let mut sys = vec![];
+        let mut user = vec![];
+
+        for fid in start_fid..=end_fid {
+            let sys_tx = build_system_transaction(&stores, fid);
+            let user_tx = build_user_transaction(&stores, fid);
+            sys.push(sys_tx);
+            user.push(user_tx);
+        }
+
+        Ok((sys, user))
+    }
+
+    pub fn handle_post_commit_message(
+        &self,
+        msg: &PostCommitMessage,
+    ) -> Result<(), ReplicationError> {
+        let block_number = match msg.header.height {
+            Some(height) => height.block_number,
+            None => {
+                return Err(ReplicationError::InvalidMessage(
+                    "PostCommitMessage must contain a block number".to_string(),
+                ));
+            }
+        };
+
+        // Clean up old snapshots
+        let expiration_block = block_number.saturating_sub(self.snapshot_options.max_age);
+        self.stores.close_snapshots_below(expiration_block);
+
+        // Check if we can take a snapshot
+        if block_number % self.snapshot_options.interval != 0 {
+            return Ok(());
+        }
+
+        // Open a snapshot
+        self.stores.open_snapshot(block_number, msg.shard_id)
+    }
+}
+
+fn collect_messages_with_page_options<
+    T: Fn(&PageOptions) -> Result<MessagesPage, crate::core::error::HubError>,
+>(
+    f: T,
+) -> Vec<proto::Message> {
+    let mut page_options = PageOptions::default();
+    let mut messages = vec![];
+
+    loop {
+        let results = f(&page_options).unwrap();
+
+        if results.messages.is_empty() {
+            break;
+        }
+
+        messages.extend(results.messages);
+
+        if results.next_page_token.is_none() {
+            break;
+        }
+
+        page_options.page_token = results.next_page_token;
+    }
+
+    messages
+}
+
+fn collect_messages<T: StoreDef + Clone>(store: &Store<T>, fid: &u64) -> Vec<proto::Message> {
+    collect_messages_with_page_options(|page_options| {
+        store.get_all_messages_by_fid(*fid, None, None, page_options)
+    })
+}
+
+fn collect_compact_state<T: StoreDef + Clone>(store: &Store<T>, fid: &u64) -> Vec<proto::Message> {
+    collect_messages_with_page_options(|page_options| {
+        store.get_compact_state_messages_by_fid(*fid, page_options)
+    })
+}
+
+fn build_user_messages_for_fid(stores: &Stores, fid: u64) -> Vec<proto::Message> {
```

Comment:
How are we handling the ordering problem here? Verifications should be committed first, then username proofs, then everything else right?

Reply:
Added to our follow-ups list